Implemented functionality for client to reserve item

harshitagirase opened the pull request.

harshitagirase committed:
Base files added

harshitagirase committed:
Base files added

harshitagirase committed:
Added nav bar to the base template

harshitagirase committed:
Added nav bar to the base template

harshitagirase committed:
Added base and issue content to files

harshitagirase committed:
Added base and issue content to files

harshitagirase committed:
Add users

harshitagirase committed:
Add users

harshitagirase committed:
Views file changed

harshitagirase committed:
Views file changed

harshitagirase committed:
Deleted unnecessary code

harshitagirase committed:
Deleted unnecessary code

harshitagirase committed:
Redirected to view

harshitagirase committed:
Redirected to view

harshitagirase committed:
S

Some BS test cases

harshitagirase committed:
S

Some BS test cases

harshitagirase committed:
Deleted duplicate test

harshitagirase committed:
Deleted duplicate test

harshitagirase committed:
Deleted random def

harshitagirase committed:
Deleted random def

harshitagirase committed:
Renamed test cases

harshitagirase committed:
Renamed test cases

harshitagirase committed:
Renamed view

harshitagirase committed:
Renamed view

harshitagirase committed:
Changed URL

harshitagirase committed:
Changed URL

harshitagirase committed:
Changed name

harshitagirase committed:
Changed name

harshitagirase committed:
Added view

harshitagirase committed:
Urls.py file added

harshitagirase committed:
Urls.py file added

harshitagirase committed:
Views modified, included inputs

harshitagirase committed:
Views modified, included inputs

harshitagirase committed:
URL modified

harshitagirase committed:
URL modified

harshitagirase committed:
Deleted base.html file

harshitagirase committed:
Deleted base.html file

harshitagirase committed:
Views.py changed

harshitagirase committed:
Views.py changed

harshitagirase committed:
Changes URL

harshitagirase committed:
Changes URL

harshitagirase committed:
Deleted templates

harshitagirase committed:
Deleted templates

StephenOrJames commented on dam/loans/views.py:
```diff
@@ -1,3 +1,34 @@
+from django.http import Http404
 # Create your views here.
+
+class validForm(forms.Form):
```
Create a `forms.py` with this form (and use a more descriptive name - perhaps `ReserveItemForm` of something of the sort).

harshitagirase replied:
What do you mean by `with this form`

StephenOrJames commented on dam/loans/views.py:
```diff
@@ -1,3 +1,34 @@
+                return redirect('/details',input_id=item_id)
+            else:
+                messages.error(request, 'Your item was not reserved. Please go back and reserve the item again.'
+                return redirect('/inventory')
```
`redirect(reverse('inventory:index'))` (you'll probably need to import `reverse` from `django.urls`).

StephenOrJames commented on dam/loans/views.py:
```diff
@@ -1,3 +1,34 @@
+    if request.method == 'POST':
+        form= validForm(request.POST)
+        if form.is_valid():
+            if(Item.objects.with_availability().get(id=item_id).available() > 0:
```
You should probably separate getting the item from checking if it is available.

Maybe try something like this:
```
try:
    item = Item.objects.with_availability().get(id=item_id)
except Item.DoesNotExist:
    raise Http404()
if item.available > 0:
    # ...
```

The problem is that if no `Item` with the ID exists, what you have would cause an internal server error since `Item.DoesNotExist` would be raised and would cause an internal server error (500).

StephenOrJames commented on dam/loans/tests.py:
```diff
@@ -1,3 +1,48 @@
 from django.test import TestCase
+from django.test import Client 
```
Unused?

StephenOrJames commented on dam/loans/urls.py:
```diff
@@ -0,0 +1,8 @@
+app_name = 'loans'
+
+urlpatterns = [
+    path('/<int:item_id>',checkIfItemAvailable, name='issue')
```
```suggestion
    path('/reserve/<int:item_id>/',checkIfItemAvailable, name='reserve')
```

StephenOrJames commented on dam/loans/urls.py:
```diff
@@ -0,0 +1,8 @@
+from django.urls import path
+from .views import *
```
Please don't do this 😢

harshitagirase replied:
Sorry :'(

harshitagirase replied:
I'm ashamed

StephenOrJames commented on dam/loans/views.py:
```diff
@@ -1,3 +1,34 @@
+    first_name = forms.CharField()
+    last_name = forms.CharField()
+    
+def valid_email(self):
```
This should be intended within the form (it's a method, not a stand-alone function). Also, the name should be `clean_<field-name>` (in this case `clean_email`) if you want Django to automatically call it.

StephenOrJames commented on dam/loans/tests.py:
```diff
@@ -1,3 +1,48 @@
+class ValidateTestCases(TestCase):
+
+    def test_details(self):
+        response= self.client.post('/inventory/details/2')
```
You want the URL path for _your_ view.

StephenOrJames committed:
changed path

Co-Authored-By: harshitagirase <harshitagirase97@gmail.com>

StephenOrJames committed:
changed path

Co-Authored-By: harshitagirase <harshitagirase97@gmail.com>

harshitagirase committed:
New forms file added

harshitagirase committed:
New forms file added

harshitagirase committed:
Changes made, resolved conflicts in urls.py and views.py

harshitagirase committed:
Resolved conflicts in urls.py and views.py

StephenOrJames commented on dam/loans/forms.py:
```diff
@@ -0,0 +1,14 @@
+    def clean_email(self):
+        email=self.cleaned_data['email']
+        if not email.endswith('@buffalo.edu'):
+            raise ValidationError('Invalid email! Please use your @buffalo.edu email address to reserve items.')
```
```suggestion
            raise forms.ValidationError('Invalid email! Please use your @buffalo.edu email address to reserve items.')
```

StephenOrJames commented on dam/loans/views.py:
```diff
@@ -18,17 +20,45 @@ def reservations(request, reservation_id):
+                return redirect('/details',input_id=item_id)
+            else:
+                messages.error(request, 'Your item was not reserved. Please go back and reserve the item again.'
+                return redirect(reverse('/inventory:index'))
```
```suggestion
                return redirect(reverse('inventory:index'))
```

StephenOrJames committed:
fixed directory name

Co-Authored-By: harshitagirase <harshitagirase97@gmail.com>

StephenOrJames committed:
imported ValidationError

Co-Authored-By: harshitagirase <harshitagirase97@gmail.com>

harshitagirase committed:
Forms changed

harshitagirase committed:
added validationerror package
Merge branch 'clientissue/final' of https://github.com/anandimous/DAM into clientissue/final

harshitagirase committed:
duplicate import code deleted

harshitagirase committed:
Removed unnecessary tests

harshitagirase committed:
Closed braces

harshitagirase committed:
Error

StephenOrJames commented on dam/loans/views.py:
```diff
@@ -18,17 +19,45 @@ def reservations(request, reservation_id):
+            except Item.DoesNotExist:
+                raise Http404()
+            if item.available> 0:
+                messages.success(request, 'Your item has been reserved! You can pick it up from Baldy 19')
```
Don't forget to actually create the `ItemReservation`!

harshitagirase committed:
ItemReservation created

harshitagirase committed:
Client and Item Reserved

anandimous committed:
Merge branch 'develop' into clientissue/final

anandimous reviewed (approved):
lgtm
